bpo-35859: re module, fix three bugs about capturing groups

animalize opened the pull request:
Fix wrong capturing groups in rare cases, the span of capturing group may be lost when backtracking.
These bugs exist since Python 2.

* macro `MARK_PUSH(lastmark)` didn't protect `MARK 0` if it was the only available mark
* jump `JUMP_MIN_UNTIL_3` needs `LASTMARK_SAVE()` and `MARK_PUSH()`
* jump `JUMP_ASSERT_NOT`  needs `LASTMARK_SAVE()` and `MARK_PUSH()`


Please read review guide in issue35859:
<!-- issue-number: [bpo-35859](https://bugs.python.org/issue35859) -->
https://bugs.python.org/issue35859
<!-- /issue-number -->

serhiy-storchaka commented on Lib/test/test_re.py:
```diff
@@ -2101,6 +2101,14 @@ def test_bug_34294(self):
+    def test_bug_35859(self):
+        # Capture behavior depends on the order of an alternation
+        s = 'ab'
+        self.assertEqual(re.search(r'(ab|a)*?b', s).groups(), ('a',))
```
Why `search()` is used instead of `match()` or `fullmatch()`?

`ab|a` is equivalent to `ab?`. Is there a reason why use the former? If there is a difference, it is better to use `.b|a` instead, because `ab|a` can be transformed to `ab?` by the RE compiler in future versions.

davisjam replied:
@serhiy-storchaka 

> Why `search()` is used instead of `match()` or `fullmatch()`?
> `ab|a` is equivalent to `ab?`. Is there a reason why use the former?

"Because it can be". `(ab|a)*?b` is a reduced test case for a real regex I found during my research.

serhiy-storchaka commented on Lib/test/test_re.py:
```diff
@@ -2101,6 +2101,14 @@ def test_bug_34294(self):
+        s = 'ab'
+        self.assertEqual(re.search(r'(ab|a)*?b', s).groups(), ('a',))
+        self.assertEqual(re.search(r'(ab|a)+?b', s).groups(), ('a',))
+        self.assertEqual(re.search(r'(ab|a){0,}?b', s).groups(), ('a',))
```
`X{0,}?` is equivalent to `X*?`, so this test is redundant.

wjssz committed:
Step 1/4, test-cases

Show the wrong behaviors before this fix.

wjssz committed:
Step 2/4, MARK_PUSH(lastmark) macro bug

MARK_PUSH(lastmark) macro didn't protect MARK-0 if it was the only available mark.

wjssz committed:
Step 3/4, JUMP_MIN_UNTIL_3 needs LASTMARK_SAVE() and MARK_PUSH()

wjssz committed:
Step 4/4, JUMP_ASSERT_NOT needs LASTMARK_SAVE()

wjssz committed:
Fix for Step 4, JUMP_ASSERT_NOT needs MARK_PUSH() as well

before this fix, the test-case returns ('c', 'c', 'c')

No need to save state->repeat in ctx->u.rep.
SRE_OP_BRANCH saves state->repeat in ctx->u.rep, this is because after JUMP_BRANCH, the state->repeat may be NULL.
While SRE_OP_ASSERT_NOT doesn't have this problem.
